src/optigraph.jl

```diff
@@ -1,745 +1,1118 @@
+    _track_node_in_graph(graph, node)
+    return nothing
+end
+
+function _track_node_in_graph(graph::OptiGraph, node::OptiNode)
+    source = source_graph(node)
+    if haskey(source.node_to_graphs, node)
+        push!(source.node_to_graphs[node], graph)
+    else
+        source.node_to_graphs[node] = [graph]
+    end
+    return nothing
+end
+
 """
-    subgraph(graph::OptiGraph, idx::Int64)
+    get_node(graph::OptiGraph, idx::Int)
 
-Retrieve the the subgraph in `graph` at index `idx`.
+Retrieve the optinode in `graph` at the given index.
 """
-subgraph(graph::OptiGraph, idx::Int64) = graph.subgraphs[idx]
-@deprecate getsubgraph subgraph
+function get_node(graph::OptiGraph, idx::Int)
+    return collect(graph.optinodes)[idx]
+end
+
 """
-    all_subgraphs(graph::OptiGraph)::Vector{OptiGraph}
+    collect_nodes(jump_func::T where T <: JuMP.AbstractJuMPScalar)
 
-Retrieve all of the contained subgraphs of `graph`, including nested subgraphs. The order of the subgraphs in
-the returned vector starts with the local subgraphs in `optigraph` and then appends the nested subgraphs for each local subgraph.
+Retrieve the optinodes contained in a JuMP expression.
 """
-function all_subgraphs(graph::OptiGraph)
-    subgraphs = graph.subgraphs
-    for subgraph in subgraphs
-        subgraphs = [subgraphs; all_subgraphs(subgraph)]
-    end
-    return subgraphs
+function collect_nodes(jump_func::T where {T<:JuMP.AbstractJuMPScalar})
+    vars = _extract_variables(jump_func)
+    nodes = JuMP.owner_model.(vars)
+    return collect(nodes)
 end
 
 """
-    subgraph_by_index(graph::OptiGraph, index::Int64)::OptiGraph
+    local_nodes(graph::OptiGraph)::Vector{<:OptiNode}
 
-Recursively search optigraph `graph` for the subngraph at `index` by traversing subgraphs.
-Note that the subgraph is not unique to the index. Since the search is depth-first, the subgraph
-returned may be different if the overall `graph` structure changes.
+Retrieve the optinodes defined within the optigraph `graph`. This does 
+not return nodes that exist in subgraphs.
 """
-subgraph_by_index(graph::OptiGraph, index::Int64) = all_subgraphs(graph)[index]
+function local_nodes(graph::OptiGraph)
+    return collect(graph.optinodes)
+end
+
+@deprecate get_nodes local_nodes
 
 """
-    num_subgraphs(graph::OptiGraph)::Int64
+    num_local_nodes(graph::OptiGraph)::Int
 
-Retrieve the number of local subgraphs in `graph`.
+Return the number of local nodes in the optigraph `graph`.
 """
-num_subgraphs(optigraph::OptiGraph) = length(optigraph.subgraphs)
+function num_local_nodes(graph::OptiGraph)
+    return length(graph.optinodes)
+end
 
 """
-    num_all_subgraphs(graph::OptiGraph)::Int64
+    all_nodes(graph::OptiGraph)::Vector{<:OptiNode}
+
+Recursively collect all optinodes in `graph` by traversing each of its subgraphs.
+"""
+function all_nodes(graph::OptiGraph)
+    nodes = collect(graph.optinodes)
+    for subgraph in graph.subgraphs
+        nodes = [nodes; all_nodes(subgraph)]
```

Comment:
I did not realize you can call functions recursively. Julia can handle this okay? You can call the same function within itself?

Reply:
yep, as long you write the recursion in a valid way. this is how we implement all of the `all_` methods.